src/Access/AccessRights.cpp

```diff
@@ -411,88 +513,47 @@ struct AccessRights::Node
-        ///
-        /// SELECT ON *.* TO user1;
-        /// REVOKE SELECT ON system.* FROM user1;
-        /// REVOKE SELECT ON mydb.* FROM user1;
-        ///
-        /// And the other node has the rights:
-        ///
-        /// SELECT ON *.* TO user2;
-        /// REVOKE SELECT ON system.* FROM user2;
-        ///
-        /// First, we check that each child from the other node is present in the current node:
-        ///
-        /// SELECT ON *.* TO user1;  -- checked
-        /// REVOKE SELECT ON system.* FROM user1; -- checked
-        if (other.children)
-        {
-            for (const auto & [name, node] : *other.children)
-            {
-                const auto & child = tryGetChild(name);
-                if (child == nullptr)
-                {
-                    if (!flags.contains(node.flags))
-                        return false;
-                }
-                else
-                {
-                    if (!child->contains(node))
-                        return false;
-                }
-            }
-        }
-
-        if (!children)
-            return true;
-
-        /// Then we check that each of our children has no other rights revoked.
-        ///
-        /// REVOKE SELECT ON mydb.* FROM user1; -- check failed, returning false
-        for (const auto & [name, node] : *children)
-        {
-            if (other.children && other.children->contains(name))
-                continue;
-
-            if (!node.flags.contains(other.flags))
-                return false;
-        }
-
-        return true;
+        Node tmp_node = *this;
+        tmp_node.makeIntersection(other);
+        /// If we get the same node after the intersection, our node is fully covered by the given one.
+        return tmp_node == other;
     }
 
     void makeUnion(const Node & other)
     {
-        makeUnionRec(other);
+        Node result;
+        Node rhs = other;
+        makeUnionRec(result, rhs);
+        children = std::move(result.children);
+        flags |= other.flags;
         optimizeTree();
     }
 
     void makeIntersection(const Node & other)
     {
-        makeIntersectionRec(other);
+        Node result;
+        Node rhs = other;
+        makeIntersectionRec(result, rhs);
+        children = std::move(result.children);
+        flags &= other.flags;
         optimizeTree();
     }
 
     ProtoElements getElements() const
     {
         ProtoElements res;
-        getElementsRec(res, {}, *this, {});
+        getElementsRec(res, {}, *this, {}, "");
         return res;
     }
 
-    static ProtoElements getElements(const Node * node, const Node * node_with_grant_option)
+    static ProtoElements getElements(const Node & node, const Node & node_with_grant_option)
     {
         ProtoElements res;
-        getElementsRec(res, {}, node, {}, node_with_grant_option, {});
+        Node tmp_node = node;
+        Node tmp_node_go = node_with_grant_option;
```

Comment:
Why do you need to copy `node` and `node_with_grant_option`?

Reply:
Yes, because we need to use the `getLeaf` function, which can't be marked as const because it can modify the tree structure.